### PR TITLE
Addressed issues raised in #44286. (`OccupiedEntry::replace_entry`)

### PR DESCRIPTION
**Pull request by Binero**
This commit renames the `replace` function to `replace_entry`, and
creates a seperate `replace_key` function for `OccupiedEntry`. The
original `replace` function did not solve the use-case where the
key needed to be replaced, but not the value. Documentation and
naming has also been updated to better reflect what the original
replace function does.

### CODE REVIEW

File: src/libstd/collections/hash/map.rs [hunk 1]
```diff
@@ -2238,35 +2238,57 @@ impl<'a, K, V> OccupiedEntry<'a, K, V> {
...
+    ///
+    /// if let Entry::Occupied(entry) = map.entry("poneyland".to_string()) {
+    ///     let old_key = entry.replace_key();
+    ///     assert_eq!(old_key, "poneyland");
+    /// }
+    /// ```
+    #[unstable(feature = "map_entry_replace", issue = "44286")]
+    pub fn replace_key(mut self) -> K {
+	let (old_key, _) = self.elem.read_mut();
+	mem::replace(old_key, self.key.unwrap())
```
Comment (kennytm) [inline src/libstd/collections/hash/map.rs]: Please change the tabs into spaces.

```
[00:04:04] tidy error: /checkout/src/libstd/collections/hash/map.rs:2289: tab character
[00:04:04] tidy error: /checkout/src/libstd/collections/hash/map.rs:2290: tab character
```

**Review by kennytm — CHANGES_REQUESTED** [review]

**Review by kennytm — APPROVED** [review]

**Comment by aidanhs** [comment]
Ping @dtolnay for review! (also pinged on IRC)

**Review by dtolnay — COMMENTED** [review]
Thanks for the reminder!

I am on board with the idea, and the implementation looks good. But could you change the usage example to illustrate a more realistic use case? Replacing "poneyland".to_string() with "poneyland".to_string() is not meaningful.

In general, usage examples are often not about *how* to use a particular API. If the user got far enough with Rust to be looking at OccupiedEntry documentation, we can assume they understand *how* to match on an enum and invoke functions. Here and elsewhere, the role of the example should be about *why* someone would want this functionality.

Either here or in a following PR, please also update the replace_entry documentation with a *why* example.

**Comment by carols10cents** [comment]
Triage ping @Binero! Does @dtolnay's request about the documentation example make sense?

**Comment by Binero** [comment]
@dtolnay @Carols10cents Oh, whoops. I missed this one. It sure does.

As for `replace_key`, I am not sure what the use case for that would be, @stepancheg requested it. `replace_entry` however is the expected behaviour when you use the entry API to replace a value, as otherwise the new key is lost, or the old key does not match the new value. It plugs the API hole resulting from the `Entry` taking ownership.

Discussion: https://github.com/rust-lang/rust/issues/44286

**Comment by stepancheg** [comment]
It can be used for example to reclaim memory used by keys. E. g.

```
struct RcString { ... }

Vec<RcString> known_strings;

fn reclaim_memory_used_by_keys(map: &mut HashMap<RcString, Something>) {
    for s in &known_strings {
        if let Entry::Occupied(e) = Entrymap.entry(s) {
            e.replace_key();
        }
    }
}
```

**Comment by shepmaster** [comment]
Triage ping @Binero! It looks like you got some feedback on the rationale for `replace_key` — will you have time to work that back into the documentation?

**Comment by Binero** [comment]
Added more examples, and rebased them all.

**Review by dtolnay — APPROVED** [review]

**Comment by dtolnay** [comment]
```
tidy error: C:\projects\rust\src\libstd\collections\hash\map.rs:2256: trailing whitespace
tidy error: C:\projects\rust\src\libstd\collections\hash\map.rs:2284: trailing whitespace
tidy error: C:\projects\rust\src\libstd\collections\hash\map.rs:2286: trailing whitespace
tidy error: C:\projects\rust\src\libstd\collections\hash\map.rs:2288: trailing whitespace
some tidy checks failed
```

**Comment by Binero** [comment]
Amended the style-changes to the last commit.

**Pull request merged**